monero-harness/src/lib.rs

```diff
@@ -48,18 +48,17 @@ const WAIT_WALLET_SYNC_MILLIS: u64 = 1000;
 /// Wallet sub-account indices.
 const ACCOUNT_INDEX_PRIMARY: u32 = 0;
 
-#[derive(Debug)]
-pub struct Monero<'c> {
-    pub docker: Container<'c, Cli, image::Monero>,
-    pub monerod_rpc_port: u16,
-    pub miner_wallet_rpc_port: u16,
-    pub alice_wallet_rpc_port: u16,
-    pub bob_wallet_rpc_port: u16,
+#[derive(Copy, Clone, Debug)]
+pub struct Monero {
+    monerod_rpc_port: u16,
+    miner_wallet_rpc_port: u16,
+    alice_wallet_rpc_port: u16,
+    bob_wallet_rpc_port: u16,
 }
 
-impl<'c> Monero<'c> {
+impl<'c> Monero {
     /// Starts a new regtest monero container.
-    pub fn new(cli: &'c Cli) -> Self {
+    pub fn new(cli: &'c Cli) -> (Self, Container<'c, Cli, image::Monero>) {
         let mut rng = rand::thread_rng();
         let monerod_rpc_port: u16 = rng.gen_range(1024, u16::MAX);
```

Comment:
What is the reason why you choose these ports yourself?

If you wouldn't specify them, testcontainers (actually `docker`) would automatically find a free one for you that you can query with `Container#get_host_port`. It seems counter-intuitive that you have to pass the `Client` to this function when actually, the `Image` is eventually passed to the `Client` within `run`.

Reply:
Not to mention that the random generation can potentially fail as it might hit one that is already taken :)

Reply:
@tcharding and @bonomat know more than me about this code :D

Reply:
The docker image we are using does not automatically expose those ports. That's why we need to explicitly map them.

+1 on the ports being in used problem.

Reply:
> The docker image we are using does not automatically expose those ports. That's why we need to explicitly map them.

That is a shame :/

Reply:
Let's see what they think: https://github.com/XMRto/monero/issues/9

Reply:
I should have told the whole story: we are slightly misusing their container and are starting 3 wallets instead of only 1. That is why the ports are not exposed by the image.

Reply:
Why don't you just start three containers for that? Sounds like scaling on the wrong level that  creates more work like having to manually assign ports instead of being able to roll with the default behaviour (assuming the ports are exposed by default).

Reply:
The setup is like this: 
* 1 monerod node
* 3 monero-wallet-rpc services

All three monero-wallet-rpc services connect to the same monerod node.

Reply:
The most recent version of testcontainers [supports running containers on a network](https://github.com/testcontainers/testcontainers-rs/blob/dev/CHANGELOG.md#added), so you could actually start those in separate containers and they could still be connected to each other.

Reply:
Cool, we would have needed that a few weeks ago :)

Reply:
Why didn't you raise it on the repo then?